datalad/utils.py

```diff
@@ -1546,10 +1546,17 @@ def get_dataset_root(path):
     The root path is returned in the same absolute or relative form
     as the input argument. If no associated dataset exists, or the
     input path doesn't exist, None is returned.
+
+    If `path` is a symlink or something other than a directory, the root dataset
+    containing its parent directory will be reported.  If none can be found, and
+    a symlink at `path` is pointing to a dataset, `path` itself will be
+    reported as the root.
     """
     suffix = '.git'
-    if not isdir(path):
-        path = dirname(path)
+    altered = None
+    if op.islink(path) or not op.isdir(path):
+        altered = path
+        path = op.dirname(path)
     apath = abspath(path)
```

Comment:
may be it worked even before because `abspath` doesn't dereference symlinks anyways:
```shell
(git-annex)hopa:~/datalad[master]
$> ls -ld dataset_description.json
lrwxrwxrwx 1 yoh yoh 42 Oct 26 11:02 dataset_description.json -> openfmri/ds000001/dataset_description.json

$> readlink -f dataset_description.json
/home/yoh/datalad/openfmri/ds000001/dataset_description.json

$> python -c 'from os.path import abspath; print(abspath("dataset_description.json"))'                                      
/home/yoh/datalad/dataset_description.json
```

that is why it is handy to have unittests which actually demonstrate the problem before "fixing"  it?